Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -447,19 +499,156 @@ Joystick::~Joystick()
     SDL_HapticClose(m_haptic);
     m_haptic = nullptr;
   }
-#endif
-
-#if SDL_VERSION_ATLEAST(2, 0, 9)
-  // stop all rumble
-  SDL_JoystickRumble(m_joystick, 0, 0, 0);
-#endif
-
+  if (m_gamecontroller)
+  {
+    // stop all rumble
+    SDL_GameControllerRumble(m_gamecontroller, 0, 0, 0);
+    // close game controller
+    SDL_GameControllerClose(m_gamecontroller);
+  }
   // close joystick
   SDL_JoystickClose(m_joystick);
 }
 
-#ifdef USE_SDL_HAPTIC
-void Joystick::HapticEffect::UpdateEffect()
+std::string GameController::Motor::GetName() const
+{
+  return "Motor";
+}
+
+void GameController::Motor::SetState(ControlState state)
+{
+  Uint16 rumble = state * std::numeric_limits<Uint16>::max();
+  SDL_GameControllerRumble(m_gc, rumble, rumble, std::numeric_limits<Uint32>::max());
+}
+
+std::string GameController::MotorL::GetName() const
+{
+  return "Motor L";
+}
+
+void GameController::MotorL::SetState(ControlState state)
+{
+  Uint16 rumble = state * std::numeric_limits<Uint16>::max();
+  SDL_GameControllerRumble(m_gc, rumble, 0, std::numeric_limits<Uint32>::max());
+}
+
+std::string GameController::MotorR::GetName() const
+{
+  return "Motor R";
+}
+
+void GameController::MotorR::SetState(ControlState state)
+{
+  Uint16 rumble = state * std::numeric_limits<Uint16>::max();
+  SDL_GameControllerRumble(m_gc, 0, rumble, std::numeric_limits<Uint32>::max());
+}
+
+void InputBackend::UpdateInput(std::vector<std::weak_ptr<ciface::Core::Device>>& devices_to_remove)
+{
+  SDL_GameControllerUpdate();
+}
+
+std::string GameController::GetName() const
+{
+  return m_name;
+}
+
+std::string GameController::GetSource() const
+{
+  return "SDL";
+}
+
+int GameController::GetSDLIndex() const
+{
+  return m_sdl_index;
+}
+
+std::string GameController::Button::GetName() const
+{
+  return s_sdl_button_names[m_button];
+}
+
+std::string GameController::Axis::GetName() const
+{
+  char direction = '\0';
+  // The triggers are only positive, and must not have a sign
+  if (m_axis < 4)
+  {
+    // Respect XInput: the vertical axes (where m_axis % 2 == 1) are inverted on SDL
+    if (m_range < 0 ^ m_axis % 2 == 1)
```

Comment:
Whatever this evaluates as (even I'm not sure...), VS emits a warning for it due to unclear operator precedence.

You may just want to take the more obvious approach and do something like

```
  if (m_axis >= 4)
  {
    // The triggers are only positive, and must not have a sign
  return std::string(s_sdl_axis_names[m_axis]);
  }

  bool negative = m_range < 0;

  // Respect XInput: the vertical axes (where m_axis % 2 == 1) are inverted on SDL
  if (m_axis % 2 == 1)
    negative = !negative;

  return std::string(s_sdl_axis_names[m_axis]) + (negative ? '-' : '+');
```